src/transforms/DummyMainFunction.cpp

```diff
@@ -0,0 +1,262 @@
+/** Insert dummy main function if one does not exist */
+
+#include <string.h>
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/IR/Value.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Type.h"
+#include "llvm/IR/GlobalValue.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Demangle/Demangle.h"
+
+#include "seadsa/CompleteCallGraph.hh"
+#include "seadsa/InitializePasses.hh"
+
+//#include "seahorn/Support/SeaDebug.h"
+#include "boost/format.hpp"
+
+static llvm::cl::opt<std::string>
+EntryPoint("entry-point",
+        llvm::cl::desc ("Entry point if main does not exist"),
+        llvm::cl::init (""));
+
+namespace previrt {
+    namespace transforms {
+
+
+        using namespace llvm;
+
+
+        class DummyMainFunction : public ModulePass
+        {
+            DenseMap<const Type*, Constant*> m_ndfn;
+
+            std::vector<std::string> parseArgs(std::string str){
+                std::vector<std::string> args;
+                
+                std::string arg = "";
+
+                for(unsigned i=0;i<str.size(); i++){
+                    if(str[i] == ','){
+                        args.push_back(arg);
+                        arg = "";
+                    } else {
+                        arg += str[i];
+                    }
+                }
+                if(arg!="") args.push_back(arg);
+                return args;
+            }
+
+
+
+            Function& makeNewNondetFn (Module &m, Type &type, unsigned num, std::string prefix)
+            {
+                std::string name;
+                unsigned c = num;
+                do
+                    name = boost::str (boost::format (prefix + "%d") % (c++));
+                while (m.getNamedValue (name));
+                Function *res = dyn_cast<Function>(m.getOrInsertFunction (name, &type).getCallee());
+                assert (res);
+                return *res;
+            }
+
+            Constant* getNondetFn (Type *type, Module& M) {
+                Constant* res = m_ndfn [type];
+                if (!res) {
+                    res = &makeNewNondetFn (M, *type, m_ndfn.size (), "verifier.nondet.");
+                    m_ndfn[type] = res;
+                }
+                return res;
+            }
+
+            std::map<std::string, std::vector<std::string>> getFnMap(Module& m){
+                std::map<std::string, std::vector<std::string>> fn_map;
+
+                for(auto &F: m){
+                    std::string demangled = demangle(F.getName());
+                    demangled = demangled.substr(0,demangled.find_first_of("("));
+                    
+                    if(fn_map.find(demangled) == fn_map.end()){
```

Comment:
lines 87-90 are redundant. The `map` method `operator[]` does exactly this if there is no an entry in the map